feat: [PAYMCLOUD-408] Add Terraform module for Cosmos DB accounts

ffppa opened the pull request:
### List of changes

- Introduced a new Terraform module for managing Cosmos DB accounts.
- Added support for Mongo and SQL APIs.
- Included configuration files, outputs, usage examples, tests, and documentation.

### Motivation and context

This change was required to streamline the provisioning of Cosmos DB resources in Azure environments, ensuring consistent and efficient resource management.

### ⚠️ Attention

The module you changed is also referenced in the IDH folder?

- [x] Yes
- [ ] No

If so, please make sure to update the IDH module as well.

- [x] I have updated the IDH module

### Type of changes

- [x] Add new module
- [ ] Update existing module
- [ ] Remove existing module

### Does this introduce a breaking change?

- [ ] Yes
- [x] No

### Other information

None.

### Run checks

Useful commands to run checks on local machine

```sh
bash .utils/terraform_run_all.sh init local
pre-commit run -a
```

ffppa committed:
Add Cosmos DB account Terraform module

Introduced a new Terraform module for managing Cosmos DB accounts with support for Mongo and SQL APIs. Added configuration files, outputs, usage examples, tests, and documentation to streamline resource provisioning within Azure environments.

ffppa committed:
Enable automatic failover configuration in CosmosDB templates

Added an `enable_automatic_failover` field to CosmosDB configuration files for UAT, PROD, and DEV environments. Simplified variable definitions in Terraform files and updated references to use `module.idh_loader.idh_config.enable_automatic_failover`. Removed unused and redundant variables for a cleaner configuration.

ffppa committed:
Update action to fix inspected sub-dependencies logic

Replaced the action version to address an issue where non-modified directories were incorrectly impacting dependency

ffppa committed:
excluded from inspection. This ensures accurate dependency analysis in the static analysis workflow.

mamari90 commented on IDH/00_idh/pagopa/dev/cosmosdb_account.yml:
```diff
@@ -0,0 +1,119 @@
+cosmos_sql4:
+  enabled: true
```
@ffppa a che serve?

mamari90 commented on IDH/00_idh/pagopa/prod/cosmosdb_account.yml:
```diff
@@ -0,0 +1,106 @@
+  enable_automatic_failover: true
+  minimal_tls_version: Tls12
+  additional_geo_locations: []
+  private_endpoint_enabled: false
```
@ffppa prod e uat dovrebbero essere privati con private endpoint

mamari90 commented on IDH/00_idh/pagopa/prod/cosmosdb_account.yml:
```diff
@@ -0,0 +1,106 @@
+    max_staleness_prefix: '100000'
+  server_version: '6.0'
+  main_geo_location_zone_redundant: true
+  enable_free_tier: true
```
@ffppa non so se è corretto in prod

mamari90 commented on IDH/00_idh/pagopa/uat/cosmosdb_account.yml:
```diff
@@ -0,0 +1,118 @@
+  burst_capacity_enabled: false
+  enable_automatic_failover: false
+  minimal_tls_version: Tls12
+  additional_geo_locations: []
```
@ffppa questa property non è usata in nessun ambiente

mamari90 commented on IDH/00_idh/pagopa/prod/cosmosdb_account.yml:
```diff
@@ -0,0 +1,106 @@
+  - EnableServerless
+  offer_type: Standard
+  consistency_policy:
+    consistency_level: Strong
```
@ffppa in prod vedo che usiamo comunque la bounded staleness, per evitare di rallentare le scritture.
a strong ci siamo passati solo temporaneamente per attività di migrazione

ffppa committed:
Enable private endpoints and restrict public access for CosmosDB

Updated CosmosDB configurations by enabling `private_endpoint_enabled` while setting `public_network_access_enabled` to `false` across environments. Also adjusted consistency policy to use `BoundedStaleness` in the `prod` environment and changed default test environment to `dev`. All redundant settings were removed for improved clarity.

ffppa committed:
Enable private endpoints and restrict public access for CosmosDB

Updated CosmosDB configurations by enabling `private_endpoint_enabled` while setting `public_network_access_enabled` to `false` across environments. Also adjusted consistency policy to use `BoundedStaleness` in the `prod` environment and changed default test environment to `dev`. All redundant settings were removed for improved clarity.

ffppa committed:
Simplify CosmosDB module examples and update README structure

Replaced verbose examples with streamlined, simplified usage patterns for CosmosDB Mongo and SQL modules. Updated README to remove outdated architecture section and migrated focus to available IDH resources. These changes improve readability and usability of the documentation.

ffppa committed:
Update README examples to fix typo in prefix comment

Corrected the typo in the example comments for the "prefix" field, replacing "pagoap" with "pagoapa". This ensures consistency and clarity in the documentation.

mamari90 commented on IDH/cosmosdb_account/variables.tf:
```diff
@@ -0,0 +1,210 @@
+  }
+}
+
+variable "additional_geo_locations" {
```
@ffppa giusto per sicurezza, metterei una validation che impedisce di fare geo-repliche in dev/uat

ffppa committed:
Enable additional geo-replication configuration in CosmosDB setup only in prod.

Added support for geo-replication through the `geo_replication_allowed` flag in environment-specific configurations. Updated test modules and variables to handle environments dynamically and include additional geo-locations when enabled.

ffppa committed:
excluded from inspection. This ensures accurate dependency analysis in the static analysis workflow.

ffppa committed:
Merge branch 'main' into PAYMCLOUD-408-Add-cosmos_db_account-on-IDH

ffppa committed:
Refactor CosmosDB module and update geo-replication logic.

Reorganize files with consistent naming and add validation to ensure geo-replication is restricted in disallowed environments. Adjusted `additional_geo_locations` to conditionally apply based on environment context for better flexibility.
